src/contract/assembled_transaction.ts

```diff
@@ -453,35 +457,82 @@ export class AssembledTransaction<T> {
     return tx;
   }
 
-  simulate = async (): Promise<this> => {
-    if (!this.built) {
-      if (!this.raw) {
-        throw new Error(
-          "Transaction has not yet been assembled; " +
-          "call `AssembledTransaction.build` first.",
-        );
-      }
+  private static async buildFootprintRestoreTransaction<T>(
+    options: AssembledTransactionOptions<T>,
+    sorobanData: SorobanDataBuilder | xdr.SorobanTransactionData,
+    account: Account,
+    fee: string
+  ): Promise<AssembledTransaction<T>> {
+    const tx = new AssembledTransaction(options);
+    tx.raw = new TransactionBuilder(account, {
+      fee,
+      networkPassphrase: options.networkPassphrase,
+    })
+      .setSorobanData(sorobanData instanceof SorobanDataBuilder ? sorobanData.build() : sorobanData)
+      .addOperation(Operation.restoreFootprint({}))
+      .setTimeout(options.timeoutInSeconds ?? DEFAULT_TIMEOUT);
+    await tx.simulate({ restore: false });
+    return tx;
+  }
 
-      this.built = this.raw.build();
+  simulate = async ({ restore }: {restore?: boolean} = {}): Promise<this> => {
+    if (!this.raw) {
+      throw new Error(
+        "Transaction has not yet been assembled; " +
+        "call `AssembledTransaction.build` first."
+      );
     }
-    this.simulation = await this.server.simulateTransaction(this.built);
+
+    restore = restore ?? this.options.restore;
+    this.built = this.raw.build();
 
     // need to force re-calculation of simulationData for new simulation
     delete this.simulationResult;
     delete this.simulationTransactionData;
+    this.simulation = await this.server.simulateTransaction(this.built);
+
+    if (Api.isSimulationRestore(this.simulation) && restore) {
```

Comment:
checking `restore` first is insignificantly cheaper :p

Reply:
insignificantly? do you want it switched?

Reply:
If you're in there might as well swap em but nbd